### PR TITLE
ci: Annotate each build with its docker image tags

### PR DESCRIPTION
**Pull request by antifuchs**
This PR adjusts the ci builder code to write the docker tags that it generates / uploads to dockerhub into an annotation on the build page.

### Motivation

This PR adds a feature that has not yet been specified

I would like to easily identify the docker tag that I need in order to spin up a container that can be tested (e.g. on mzcloud); the current way to do it, as far as I can tell, is to dig through a bunch of scrollback until you find https://buildkite.com/materialize/tests/builds/28518#907d38b3-4f7d-4210-b10b-1788aa6bb605

### Tips for reviewer

This PR's builds should in fact have an annotation! ...aaaand it does: https://buildkite.com/materialize/tests/builds/28606#annotation-default

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Comment by benesch** [comment]
Oh, btw, this is just for humans right? If so would highly recommend using an annotation instead of an artifact! https://buildkite.com/docs/agent/v3/cli-annotate

**Comment by antifuchs** [comment]
Hah! Annotations are in fact exactly what I may be looking for. I'll go and rework it with those (and put their creation in the right spot).

File: misc/python/materialize/mzbuild.py [hunk 1]
```diff
@@ -683,6 +684,21 @@ def ensure(self) -> None:
         for dep in self:
             dep.ensure()
 
+    def annotate_buildkite_with_tags(self) -> None:
+        """If running in buildkite, annotate the build with each dependency's docer tag."""
+        if os.environ["BUILDKITE"] is None:
+            # Not running in bk, bail.
+            return
```
[the remark this answers is not in the document]
Comment (antifuchs) [reply misc/python/materialize/mzbuild.py]: On it. Fixing the formatting, too (:

File: misc/python/materialize/mzbuild.py [hunk 2]
```diff
@@ -534,7 +534,8 @@ def acquire(self) -> None:
     def ensure(self) -> None:
         """Ensure the image exists on Docker Hub if it is publishable.
 
-        The image is pushed using its spec as its tag.
+        The image is pushed using its spec as its tag. Its tag is written
```
Comment (benesch) [inline misc/python/materialize/mzbuild.py:537]: minor, but this comment is stale now!

**Review by benesch — APPROVED** [review]
So good! 🙌🏽 

(I mean, I think. I guess we have to wait another minute to see what it actually looks like.)

**Comment by benesch** [comment]
Hm. I guess Buildkite doesn't like the HTML?

image

**Comment by benesch** [comment]
Hmm, but the test summary result plugin totally uses those elements: https://github.com/buildkite-plugins/junit-annotate-buildkite-plugin/blob/6935f26aa53508f1b55b34834583056dfbd34908/tests/command.bats

**Comment by benesch** [comment]
> CommonMark supports HTML inside Markdown blocks, but will revert to Markdown parsing on newlines. For more information about how HTML is parsed and which tags CommonMark supports please refer to the [CommonMark spec](https://spec.commonmark.org/).

Hm.

**Comment by antifuchs** [comment]
Haha, is it the newlines?! I’m also fairly sure there may be a problem with indentation or the first line. Maybe it is a commonmark/normal md issue.

**Comment by benesch** [comment]
Yeah, sure seems like the `dedent` isn't getting applied. Weird. Maybe try the thing the `textwrap.dedent` docs suggest to avoid the first blank line? https://docs.python.org/3/library/textwrap.html#textwrap.dedent

Or just ditch `dedent` and let the source code be a little ugly?

**Comment by antifuchs** [comment]
ALMOST

**Comment by antifuchs** [comment]
OK, think this looks ready now: https://buildkite.com/materialize/tests/builds/28613#annotation-default is the current build's annotation, and it seems to pass the tests that apply to it.

**Review by benesch — APPROVED** [review]
Looks beautiful! LGTM. You're based on a branch that has a flaky `cargo test` job, though. You can either retry that job a few times (empirically takes about three tries), or rebase on the latest main.

**Comment by benesch** [comment]
Addressed the last little nit above and marked for automerge!

**Pull request merged**